### PR TITLE
Fix accessible object inherritance when used with GTK >= 3.2

### PR DESCRIPTION
**Pull request by TheMuso**
Also disable eel accessibility code that is not needed.

Fixes https://github.com/mate-desktop/caja/issues/590

### CODE REVIEW

**Comment by raveit65** [comment]
Thanks, you can use 3.0.0 as Version check, 3.14 is our minimal supported version.
With 1.18 we make caja gtk3 only, so every version check < 3.0.0 will be normal code.
I will test this over the weekend.

**Comment by TheMuso** [comment]
On Sat, Oct 08, 2016 at 05:40:34PM AEDT, raveit65 wrote:
> Thanks, you can use 3.0.0 as Version check, 3.14 is our minimal supported version.
> With 1.18 we make caja gtk3 only, so every version check < 3.0.0 will be normal code.
> I will test this over the weekend.

Fair enough, was going based on when the accessibility code changes were made 
in GTK 3. If you would prefer, I can change it.

Luke

**Comment by raveit65** [comment]
Caja desktop is speaking now, same for caja window........dunno if this was broken.
Thank you.
> If you would prefer, I can change it.

yes please if you're familiar with 'git rebase -i'.
@monsta 
Can you do a code review please?

**Comment by TheMuso** [comment]
On Mon, Oct 10, 2016 at 07:29:39PM AEDT, raveit65 wrote:
> Caja desktop is speaking now, same for caja window........dunno if this was broken.
> Thank you.
> > If you would prefer, I can change it.
> 
> yes please if you're familiar with 'git rebase -i'.

Ok, done, and rebased on latest master.

Luke

File: eel/eel-labeled-image.c [hunk 1]
```diff
@@ -2312,7 +2323,32 @@ static void
 eel_labeled_image_accessible_initialize (AtkObject *accessible,
         gpointer   widget)
 {
+#if GTK_CHECK_VERSION(3, 2, 0)
+    a11y_parent_class->initialize (accessible, widget);
+
+    if (GTK_IS_CHECK_BUTTON (widget))
```
Comment (monsta) [inline eel/eel-labeled-image.c]: Are these changes needed for this function? I don't see that in other classes.

**Comment by monsta** [comment]
In GTK+3 build I've got this after opening Emblems & Colors dialog:
```
(caja:28013): Gtk-CRITICAL **: gtk_widget_class_set_accessible_type: assertion 'g_type_is_a (type, widget_class->priv->accessible_type)' failed
```
Happens only once per Caja instance, need to restart it and open that dialog again to get it.
Well, that dialog doesn't really work in GTK+3 (https://github.com/mate-desktop/caja/issues/506), but I thought I'd notify you anyway.

**Comment by TheMuso** [comment]
On Mon, Oct 10, 2016 at 11:53:46PM AEDT, monsta wrote:
> monsta commented on this pull request.
> 
> 
> 
> > @@ -2312,7 +2323,32 @@ static void
>  eel_labeled_image_accessible_initialize (AtkObject *accessible,
>          gpointer   widget)
>  {
> +#if GTK_CHECK_VERSION(3, 2, 0)
> +    a11y_parent_class->initialize (accessible, widget);
> +
> +    if (GTK_IS_CHECK_BUTTON (widget))
> 
> Are these changes needed for this function? I don't see that in other classes.

Yes, it is to implement setting atk role, as is done in eel_imaged_label_accessible_get_type for the GTK 2 code path. The atk role is important, as it tells the assistive technology how it should handle working with the object in question.

**Comment by TheMuso** [comment]
On Mon, Oct 10, 2016 at 11:57:56PM AEDT, monsta wrote:
> monsta commented on this pull request.
> 
> 
> 
> > @@ -68,6 +68,9 @@
>  #include <stdio.h>
>  #include <gdk/gdkprivate.h>
>  #include <gtk/gtk.h>
> +#if GTK_CHECK_VERSION(3,2,0)
> 
> Looks like it's already included in .h file.

If you mean gtk/gtk-a11y.h, I don't think gtk.h includes it. If this is not what you are referring to, could you explain in a little more detail please?

Thanks.

**Comment by TheMuso** [comment]
Correction, I meant to say that those changes are needed to do the same as is done in eel_labeled_image_get_accessible.

**Comment by TheMuso** [comment]
On Tue, Oct 11, 2016 at 01:55:28AM AEDT, monsta wrote:
> In GTK+3 build I've got this after opening Emblems & Colors dialog:
> ```
> (caja:28013): Gtk-CRITICAL **: gtk_widget_class_set_accessible_type: assertion 'g_type_is_a (type, widget_class->priv->accessible_type)' failed
> ```
> Happens only once per Caja instance, need to restart it and open that dialog again to get it.
> Well, that dialog doesn't really work in GTK+3 (https://github.com/mate-desktop/caja/issues/506), but I thought I'd notify you anyway.

Ok thanks. I am actually wondering whether it may be eel-labeled-image. When working on that, I was actually pondering creating new objects for the various labeled image button subtypes that associate with EelLabeledImageButton, but when I saw that only the atk role was different, I decided not to go that route. Now I am wondering whether that may fix this assertion failure.

Leave it with me, I'll see what I can do.

**Comment by monsta** [comment]
> If you mean gtk/gtk-a11y.h, I don't think gtk.h includes it. If this is not what you are referring to, could you explain in a little more detail please?

No, I meant ```eel/eel-canvas.h```.

**Comment by monsta** [comment]
> Correction, I meant to say that those changes are needed to do the same as is done in eel_labeled_image_get_accessible.

Ah, right, missed that function...

**Comment by TheMuso** [comment]
On Mon, Oct 10, 2016 at 11:53:46PM AEDT, monsta wrote:
> monsta commented on this pull request.
> 
> 
> 
> > @@ -2312,7 +2323,32 @@ static void
>  eel_labeled_image_accessible_initialize (AtkObject *accessible,
>          gpointer   widget)
>  {
> +#if GTK_CHECK_VERSION(3, 2, 0)
> +    a11y_parent_class->initialize (accessible, widget);
> +
> +    if (GTK_IS_CHECK_BUTTON (widget))
> 
So to solve the assertion, I realised that these are not actually needed, so as of the latest commit in my branch, this is gone, and the assertion is also gone.

**Comment by monsta** [comment]
Ok. Is ```eel_labeled_image_accessible_get_type``` function needed now? I see that it's declared, but I can't find the function body anywhere.

I'm asking because there are some similarly named static functions that are defined later in the code:
```C
/* derived types so we can add our accessibility interfaces */
static GType         eel_labeled_image_button_get_type        (void);
static GType         eel_labeled_image_check_button_get_type  (void);
static GType         eel_labeled_image_radio_button_get_type  (void);
static GType         eel_labeled_image_toggle_button_get_type (void);
```

**Comment by TheMuso** [comment]
On Tue, Oct 11, 2016 at 09:29:51PM AEDT, monsta wrote:
> Ok. Is ```eel_labeled_image_accessible_get_type``` function needed now? I see that it's declared, but I can't find the function body anywhere.

The eel_labeled_image_accessible_get_type method is created by the G_DEFINE_TYPE_WITH_CODE macro.

Luke

**Comment by monsta** [comment]
Ok, I didn't realize that.

@raveit65 alright, you can merge it if Orca works for you.

**Comment by raveit65** [comment]
I just tested it again.
caja desktop- and browser-window are speaking well.
Thanks a lot for this PR.
